src/infer.ts

```diff
@@ -305,7 +408,36 @@ function get_object_type_info(root_file: string, checker: ts.TypeChecker, ty: an
   return null;
 }
 
-export function programInfo(filename: string, vendorPath: string, perform_vendor: boolean): ProgramInfo {
+function unwrap_nullable_type(ty: ts.Type): [ts.Type, NullOrUndefinability] | null {
+  if (!ty.isUnion()) return null;
+
+  const isNullable = ty.types.find(is_null_type) !== undefined;
+  const isUndefined = ty.types.find(is_undefined_type) !== undefined;
+  const nullOrUndefinability =
+    isNullable
+      ? isUndefined
+        ? NullOrUndefinability.AcceptsEither
+        : NullOrUndefinability.AcceptsNullOnly
+      : isUndefined
+        ? NullOrUndefinability.AcceptsUndefinedOnly
+        : null;
```

Comment:
Would you prefer mutation of a variable inside an if/else if? 🤮 (The correct answer is 'no' 😜)

Reply:
Yeah not really any better alternatives. Nested ternary's always raise an eyebrow though. Maybe just parenthesise it a bit so that the precedence is crystal clear?

Reply:
Done.